### PR TITLE
moved part of the code that is using objectMapper to Codable

### PR DESCRIPTION
**Pull request by Veleal**
this is only 1 stage of moving to Codable. Some part of the code may be changed/optimised on stage 2

### CODE REVIEW

File: FitpaySDK/A2AVerification/A2AContext.swift [hunk 1]
```diff
@@ -6,25 +6,14 @@
 //
 
 import UIKit
-import ObjectMapper
 
-open class A2AContext: NSObject, Mappable {
+open class A2AContext: NSObject, Serializable {
```
Comment (jeremiahthomasharris) [inline FitpaySDK/A2AVerification/A2AContext.swift]: Is there a reason to use Serializable over Codable?

File: FitpaySDK/Notifications/NotificationDetail.swift [hunk 2]
```diff
@@ -11,17 +10,25 @@ open class NotificationDetail : Mappable
     open var clientId: String?
     var restClient:RestClient?
 
-    public required init?(map: Map) {
+    private enum CodingKeys: String, CodingKey {
+        case ackSync = "_links.ackSync.href"
+        case completeSync = "_links.completeSync.href"
+        case type = "type"
```
Comment (jeremiahthomasharris) [inline FitpaySDK/Notifications/NotificationDetail.swift]: Any cases that match the string you can remove the ="xxx" from for cleaner code.

File: FitpaySDK/Notifications/NotificationDetail.swift [hunk 3]
```diff
@@ -11,17 +10,25 @@ open class NotificationDetail : Mappable
...
+
+    public required init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        ackSync = try container.decode(.ackSync)
+        completeSync = try container.decode(.completeSync)
+        type = try container.decode(.type)
+        syncId = try container.decode(.syncId)
+        deviceId = try container.decode(.deviceId)
+        userId = try container.decode(.userId)
+        clientId = try container.decode(.clientId)
```
Comment (jeremiahthomasharris) [inline FitpaySDK/Notifications/NotificationDetail.swift]: If we aren't doing anything special in the init from decoder we can leave it off and the compiler will generate it. This also applies to encodes.

File: FitpaySDK/Rest/Models/ApduPackage.swift [hunk 4]
```diff
@@ -52,23 +52,44 @@ open class ApduPackage : NSObject, Mappable
...
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        links = try container.decode(.links, transformer: ResourceLinkTypeTransform())
+        seIdType = try container.decode(.seIdType)
+        targetDeviceType = try container.decode(.targetDeviceType)
+        targetDeviceId = try container.decode(.targetDeviceId)
+        packageId = try container.decode(.packageId)
+        seId = try container.decode(.seId)
+        apduCommands = try container.decode(.apduCommands)
+        validUntil = try container.decode(.validUntil)
+        validUntilEpoch = try container.decode(.validUntil, transformer: CustomDateFormatTransform(formatString: "yyyy-MM-dd'T'HH:mm:ss.SSSX"))
```
Comment (jeremiahthomasharris) [inline FitpaySDK/Rest/Models/ApduPackage.swift]: You can set a date formatter on the JSONEncoder object instead.

File: FitpaySDK/A2AVerification/A2AContext.swift (hunk 1 above)
Comment (Veleal) [reply FitpaySDK/A2AVerification/A2AContext.swift]: Serializable is a custom class that inherits Codable. I added there some useful methods

File: FitpaySDK/Rest/Models/EncryptionKey.swift [hunk 5]
```diff
@@ -14,22 +12,46 @@ open class EncryptionKey: NSObject, Mappable
...
+    public required init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        links = try container.decode(.links, transformer: ResourceLinkTypeTransform())
+        keyId = try container.decode(.keyId)
+        created = try container.decode(.created)
+        if let stringNumber: String = try container.decode(.createdEpoch) {
+            createdEpoch = NSTimeIntervalTypeTransform().transform(stringNumber)
+        } else if let intNumber: Int = try container.decode(.createdEpoch) {
+            createdEpoch = NSTimeIntervalTypeTransform().transform(intNumber)
+        }
```
Comment (jeremiahthomasharris) [inline FitpaySDK/Rest/Models/EncryptionKey.swift]: Does createdEpoch not always come back as either a string or an int? Shouldn't need the if/else here. Same question for everywhere this pattern is used.

File: FitpaySDK/Extensions/Codable+JSON.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        if let stringJson = any as? String, let stringData = stringJson.data(using: .utf8) {
+            let jsonArray = try JSONSerialization.jsonObject(with: stringData, options : .allowFragments) as! [String: Any]
+            data = try JSONSerialization.data(withJSONObject: jsonArray, options: .prettyPrinted)
+        } else if let jSONObject = any {
+            let jSONObjectData = try JSONSerialization.data(withJSONObject: jSONObject, options: .prettyPrinted)
+            data = jSONObjectData
+        }
+
+        self = try JSONDecoder().decode(Self.self, from: data)
+    }
```
Comment (jeremiahthomasharris) [inline FitpaySDK/Extensions/Codable+JSON.swift]: Nice work on the init function. cleans up the code other places.

File: FitpaySDK/PaymentDevice/IdVerificationResponse.swift [hunk 7]
```diff
@@ -74,40 +73,91 @@ open class IdVerificationResponse: NSObject, Mappable {
...
-        billingLine2 <- map["billingLine2"]
-        billingCity <- map["billingCity"]
-        billingState <- map["billingState"]
-        billingZip <- map["billingZip"]
-        locale <- map["locale"]
-        
+
+    public required init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        oemAccountInfoUpdatedDate = try container.decode(.oemAccountInfoUpdatedDate, transformer: CustomDateFormatTransform(formatString: "yyyy-MM-dd'T'HH:mm:ssZZZZZ"))
```
Comment (jeremiahthomasharris) [inline FitpaySDK/PaymentDevice/IdVerificationResponse.swift]: Move date format string into a private variable so its all in one place.

File: FitpaySDK.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -1333,6 +1298,7 @@
 			isa = PBXSourcesBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+				916FC0C220A0366A0071B66D /* Codable+JSON.swift in Sources */,
```
Comment (jeremiahthomasharris) [inline FitpaySDK.xcodeproj/project.pbxproj]: Break `SerializableExtension` into its own file, `KeyedEncodingContainerExtension` into its own file and move keyedDecoding into KeyedDecodingContainerExtensions. Removing Codable+JSON

File: FitpaySDK/A2AVerification/A2AVerificationRequest.swift [hunk 9]
```diff
@@ -6,24 +6,13 @@
...
     open var context: A2AContext?
-    
-    public required init?(map: Map) {
-        super.init()
-    }
-    
+
     internal override init() {
         super.init()
     }
```
Comment (jeremiahthomasharris) [inline FitpaySDK/A2AVerification/A2AVerificationRequest.swift]: init can be removed

File: FitpaySDK/Rest/Models/EncryptionKey.swift (hunk 5 above)
Comment (climbonbelay) [reply FitpaySDK/Rest/Models/EncryptionKey.swift]: From the platform side, it will always be an int

Comment (Veleal) [reply FitpaySDK/Rest/Models/EncryptionKey.swift]: So I can remove "if let stringNumber: String = try container.decode(.createdEpoch)" part cuz it supposed to receive int value, right?

Comment (jeremiahthomasharris) [reply FitpaySDK/Rest/Models/EncryptionKey.swift]: correct, and change the line decoding it to the standard NSTimeInterval pattern you created.

**Comment by climbonbelay** [comment]
@Veleal can you check to see why `testNSTimeIntervalToInt` is failing?

**Comment by jeremiahthomasharris** [comment]
Fixed

**Pull request merged**